### PR TITLE
3754-1

### PR DESCRIPTION
**Pull request by vlasov95**
## Номер
3754

## Номер задания
1

## Ссылка на видео с демо
https://www.youtube.com/watch?v=0nrpfPka_CA&feature=youtu.be

### CODE REVIEW

File: 3754/1/top_gems.rb [hunk 1]
```diff
@@ -0,0 +1,29 @@
+require './libs/gems_finder.rb'
```
Comment (anatoliliotych) [inline 3754/1/top_gems.rb]: common name for folder is lib :)

File: 3754/1/top_gems.rb [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+require 'bundler'
+require 'yaml'
+require 'optparse'
+Bundler.require
+
+show_gems = GemsFinder.new
```
Comment (anatoliliotych) [inline 3754/1/top_gems.rb]: I'd argue about names of class

File: 3754/1/top_gems.rb [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+OptionParser.new do |opts|
+  opts.on('--file[=FILE]') do |file|
+    show_gems.get_github_url(YAML.load_file(file))
```
Comment (anatoliliotych) [inline 3754/1/top_gems.rb]: what if file is not passed?

File: 3754/1/top_gems.rb [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+  opts.on('--top[=TOP]', Integer) do |top|
+    top_array = []
+    (0..top - 1).each do |element|
+      top_array << show_gems.array_of_gems[element] unless show_gems.array_of_gems[element].nil?
+    end
+    show_gems.rewrite_final_array(top_array)
+  end
+end.parse!
+show_gems.put_into_console
```
Comment (anatoliliotych) [inline 3754/1/top_gems.rb]: In general I'd remove all logic from this file to keep as clean as possible.
So the idea is to have class which accepts params and provides final output

File: 3754/1/libs/check_github.rb [hunk 5]
```diff
@@ -0,0 +1,10 @@
+# class for checking github url
+class CheckGithub
```
Comment (anatoliliotych) [inline 3754/1/libs/check_github.rb]: class names should be like GithubChecker

File: 3754/1/libs/check_source_code_url.rb [hunk 6]
```diff
@@ -0,0 +1,14 @@
+require './libs/check_github.rb'
+# class for checking source code url
+class CheckSourceCodeURL
+  def initialize(source_code_url)
+    @source_code_url = source_code_url
+  end
+
+  def check_sc_url
+    check_git_for_sc = CheckGithub.new(@source_code_url)
```
Comment (anatoliliotych) [inline 3754/1/libs/check_source_code_url.rb]: I don't see any reason to move github check into separate class.

File: 3754/1/libs/check_url.rb [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+# class for checking homepage url
+class CheckURL
+  def initialize(homepage_url, source_code_url)
+    @homepage_url = homepage_url.to_s
+    @source_code_url = source_code_url.to_s
+  end
+
+  def check_url
+    check_source_url = CheckSourceCodeURL.new(@source_code_url)
+    check_git_for_hp = CheckGithub.new(@homepage_url)
```
Comment (anatoliliotych) [inline 3754/1/libs/check_url.rb]: you check twice for github, is there any reason to not trust to the first check? :)

File: 3754/1/libs/take_data.rb [hunk 8]
```diff
@@ -0,0 +1,38 @@
+require 'open-uri'
+# class for taking data about gem
+class TakeData
```
Comment (anatoliliotych) [inline 3754/1/libs/take_data.rb]: so you can have DataFetcher class which parse data and initialize an object like GemStat which then collected to GemStatCollection object which is like storage

**Comment by vlasov95** [comment]
@anatoliliotych I've change the structure and tried to make fixes according to your comments

File: 3754/1/lib/data_fetcher.rb [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+  attr_reader :list_of_urls, :stats
+
+  def initialize(list_of_urls)
+    @list_of_urls = list_of_urls
+    @stats = []
+  end
+
+  def collect_all_data
+    list_of_urls.each do |gems|
+      stats << DataParser.new(gems[1]).only_stats.insert(0, gems[0])
```
Comment (anatoliliotych) [inline 3754/1/lib/data_fetcher.rb]: replace list_of_urls with hash instead of array

File: 3754/1/lib/data_parser.rb [hunk 10]
```diff
@@ -0,0 +1,40 @@
+require 'open-uri'
+# Class for parsing github url of gem
+class DataParser
+  def initialize(github_url)
```
Comment (anatoliliotych) [inline 3754/1/lib/data_parser.rb]: initialize with hash with name and url

**Comment by vlasov95** [comment]
@anatoliliotych I hope it became better than it was

**Pull request merged**